### ISSUE TITLE
Clippy warns twice about unknown lints

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#6602]
I tried this code:

```rust
#[warn(x)]
fn main() {}
```

I expected to see this happen: Clippy warns once that `x` is an unknown lint.

Instead, this happened: Clippy warns twice.

```
warning: unknown lint: `x`
 --> src/main.rs:1:8
  |
1 | #[warn(x)]
  |        ^
  |
  = note: `#[warn(unknown_lints)]` on by default

warning: unknown lint: `x`
 --> src/main.rs:1:8
  |
1 | #[warn(x)]
  |        ^

warning: 2 warnings emitted
```

### Meta

- `cargo clippy -V`: 0.1.51 (2021-01-16 8a65184)
- `rustc -V`: 1.49.0

See also https://github.com/rust-lang/rust/pull/80524#discussion_r558984197 - this looks unrelated to the `UNKNOWN_CLIPPY_LINTS` pass because it shows up even with that disabled.

cc @flip1995

**Comment by flip1995** [comment]
This is not a Clippy issue, but a rustc issue in general. This happens when linting attributes. Somehow in rustc diagnostics doesn't realize, that this warning was already emitted and emits it again. rustc diagnostics has code that hashes warnings/errors, to make sure that they are not emitted twice. This code fails for attributes.

**Comment by jyn514** [comment]
@flip1995 this issue doesn't happen with rustdoc:
```
> rustdoc warn.rs 
warning: unknown lint: `x`
 --> warn.rs:1:8
  |
1 | #[warn(x)]
  |        ^
  |
  = note: `#[warn(unknown_lints)]` on by default

warning: 1 warning emitted
```
The issue may be *caused* by rustc, but it's specific to clippy.

**Comment by jyn514** [comment]
> rustc diagnostics has code that hashes warnings/errors, to make sure that they are not emitted twice, which for some reason fails for attributes (for the first attribute linted).

This may be true, but rustdoc still only emits one lint even with `-Z deduplicate-diagnostics=no`.

**Comment by flip1995** [comment]
Oh, I see. So Clippy's lint machinery somehow breaks rustc's diagnostics 🤔

**Comment by flip1995** [comment]
I label it as `E-hard`, since I already put in many hours trying to fix it about 2 years ago and wasn't even able to find the cause.
